dashboard/app/controllers/courses_controller.rb

```diff
@@ -189,6 +143,53 @@ def i18n_params
 
   private
 
+  def get_unit_group
+    course_name = params[:course_name]
+
+    unit_group =
+      params[:action] == "edit" ?
+        UnitGroup.get_without_cache(course_name) :
+        UnitGroup.get_from_cache(course_name)
+
+    return unit_group if unit_group
+
+    # When the url of a course family is requested, redirect to a specific course version.
+    if UnitGroup.family_names.include?(params[:course_name])
+      unit_group = UnitGroup.latest_stable_version(params[:course_name])
+      redirect_to action: params[:action], course_name: unit_group.name if unit_group
+    end
+
+    unit_group
+  end
+
+  def set_unit_group
+    @unit_group = get_unit_group
+    raise ActiveRecord::RecordNotFound unless @unit_group
+  end
+
+  def render_no_access
+    if params[:action] == "show" && @unit_group.plc_course
+      authorize! :show, Plc::UserCourseEnrollment
+      user_course_enrollments = [Plc::UserCourseEnrollment.find_by(user: current_user, plc_course: @unit_group.plc_course)]
+      render 'plc/user_course_enrollments/index', locals: {user_course_enrollments: user_course_enrollments}
+      return
+    end
+
+    if @unit_group.pilot?
+      authenticate_user!
+      unless @unit_group.has_pilot_access?(current_user)
+        return render :no_access
+      end
+    end
+
+    if @unit_group.in_development?
+      authenticate_user!
+      unless current_user.permission?(UserPermission::LEVELBUILDER)
+        return render :no_access
+      end
+    end
```

Comment:
the no_access checks here look like they were previously only being done on `show`, and are now also being done on `:vocab, :resources, :code, :standards`. what's the thinking behind adding these checks to this path? 

for `:vocab, :resources, :code, :standards`, I would suggest moving this logic into CanCan, and letting it return the default of `forbidden` instead of the special `no_access` page, unless we have a clear product reason for preferring `no_access`.

Reply:
if you feel strongly about no_access... this technique could make it possible to quickly override all CanCan exceptions with the friendlier page: https://github.com/code-dot-org/code-dot-org/blob/cba900f0ae44ae195462ed22b58e8c1f3d853ea4/dashboard/app/controllers/projects_controller.rb#L489

Reply:
and you can use this technique to handle different actions differently: https://github.com/code-dot-org/code-dot-org/blob/16a68867300a77739dcbc94fc567c498cffa041a/dashboard/app/controllers/api/v1/sections_controller.rb#L8-L10

Reply:
I was looking to make things more consistent across the different actions which is why I set it up this way. I am planning in the follow PR to show the no_access message if you can't instruct or participate in the course. I think I'll check with Tess since the spec does not really call out what to do here

Reply:
if consistency is not something you want to tackle in this PR... then I would leave behavior of vocab, etc alone for now. in the case you could let those return 403 via cancan.